Mathlib/GroupTheory/GroupAction/Jordan.lean

```diff
@@ -0,0 +1,673 @@
+  intro H
+  apply not_le.mpr h
+  rw [← ncard_inter_add_ncard_union, H, ncard_univ]
+  exact Nat.le_add_left (Nat.card α) (s ∩ t).ncard
+
+theorem ncard_pigeonhole_compl'
+    (h : s.ncard + t.ncard < Nat.card α) :
+    s ∪ t ≠ ⊤ := by
+  intro h'
+  apply not_le.mpr h
+  rw [← ncard_univ, ← top_eq_univ, ← h']
+  exact ncard_union_le s t
+
+end Set
+
+end PigeonHole
+
+open MulAction SubMulAction Subgroup
+
+open scoped Pointwise
+
+section Jordan
+
+variable {G α : Type*} [Group G] [MulAction G α]
+
+/-- In a 2-transitive action, the normal closure of stabilizers is the full group. -/
+theorem normalClosure_of_stabilizer_eq_top (hsn' : 2 < ENat.card α)
+    (hG' : IsMultiplyPretransitive G α 2) {a : α} :
+    normalClosure ((stabilizer G a) : Set G) = ⊤ := by
+  have _ : IsPretransitive G α := by
+    rw [← is_one_pretransitive_iff]
+    exact isMultiplyPretransitive_of_le' (one_le_two) (le_of_lt hsn')
+  have _ : Nontrivial α := by
+    rw [← ENat.one_lt_card_iff_nontrivial]
+    exact lt_trans (by norm_num) hsn'
+  have hGa : IsCoatom (stabilizer G a) :=  by
+    rw [isCoatom_stabilizer_iff_preprimitive]
+    exact isPreprimitive_of_is_two_pretransitive hG'
+  apply hGa.right
+  -- Remains to prove: (stabilizer G a) < Subgroup.normalClosure (stabilizer G a)
+  constructor
+  · apply le_normalClosure
+  · intro hyp
+    have : Nontrivial (ofStabilizer G a) := by
+      rw [← ENat.one_lt_card_iff_nontrivial]
+      apply lt_of_add_lt_add_right
+      rwa [ENat_card_ofStabilizer_add_one_eq]
+    rw [nontrivial_iff] at this
+    obtain ⟨b, c, hbc⟩ := this
+    have : IsPretransitive (stabilizer G a) (ofStabilizer G a) := by
+      rw [← is_one_pretransitive_iff]
+      rwa [← ofStabilizer.isMultiplyPretransitive]
+    -- get g ∈ stabilizer G a, g • b = c,
+    obtain ⟨⟨g, hg⟩, hgbc⟩ := exists_smul_eq (stabilizer G a) b c
+    apply hbc
+    rw [← SetLike.coe_eq_coe] at hgbc ⊢
+    obtain ⟨h, hinvab⟩ := exists_smul_eq G (b : α) a
+    rw [eq_comm, ← inv_smul_eq_iff] at hinvab
+    rw [← hgbc, SetLike.val_smul, ← hinvab, inv_smul_eq_iff, eq_comm]
+    simp only [subgroup_smul_def, smul_smul, ← mul_assoc, ← mem_stabilizer_iff]
+    exact hyp (normalClosure_normal.conj_mem g (le_normalClosure hg) h)
+
+variable [Finite α]
+
+/- -- TODO : prove :
+theorem strong_jordan_of_pretransitive (hG : IsPreprimitive G α)
+    {s : Set α} {n : ℕ } (hsn : Nat.card s = n.succ)
+    (hsn' : 1 + n.succ < fintype.card α)
+    (hs_trans : IsPretransitive (fixingSubgroup G s) (SubMulAction.ofFixingSubgroup G s)) :
+    IsMultiplyPretransitive (Subgroup.normalClosure (fixingSubgroup G s : Set G)) α 2 :=
+  sorry
+-/
+
+open MulAction.IsPreprimitive
+
+open scoped Pointwise
+
+/-- A criterion due to Jordan for being 2-pretransitive (Wielandt, 13.1) -/
+theorem is_two_pretransitive_weak_jordan [DecidableEq α]
+    (hG : IsPreprimitive G α) {s : Set α} {n : ℕ}
+    (hsn : s.ncard = n.succ) (hsn' : 1 + n.succ < Nat.card α)
```

Comment:
I think `n + 1` is preferred over `n.succ`.

Reply:
OK.